usecases/telemetry/telemetry.go

```diff
@@ -0,0 +1,211 @@
+
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/entities/models"
+	"github.com/weaviate/weaviate/entities/verbosity"
+	"github.com/weaviate/weaviate/usecases/config"
+)
+
+const (
+	defaultConsumerURL  = "https://us-central1-semi-production.cloudfunctions.net/weaviate-telemetry"
+	defaultPushInterval = 24 * time.Hour
+)
+
+type nodesStatusGetter interface {
+	LocalNodeStatus(className, output string) *models.NodeStatus
+}
+
+type modulesProvider interface {
+	GetMeta() (map[string]interface{}, error)
+}
+
+// Telemeter is responsible for managing the transmission of telemetry data
+type Telemeter struct {
+	machineID         strfmt.UUID
+	nodesStatusGetter nodesStatusGetter
+	modulesProvider   modulesProvider
+	logger            logrus.FieldLogger
+	shutdown          chan struct{}
+	failedToStart     bool
+	consumerURL       string
+	pushInterval      time.Duration
+}
+
+// New creates a new Telemeter instance
+func New(nodesStatusGetter nodesStatusGetter, modulesProvider modulesProvider,
+	logger logrus.FieldLogger,
+) *Telemeter {
+	tel := &Telemeter{
+		machineID:         strfmt.UUID(uuid.NewString()),
+		nodesStatusGetter: nodesStatusGetter,
+		modulesProvider:   modulesProvider,
+		logger:            logger,
+		shutdown:          make(chan struct{}),
+		consumerURL:       defaultConsumerURL,
+		pushInterval:      defaultPushInterval,
+	}
+	return tel
+}
+
+// Start begins telemetry for the node
+func (tel *Telemeter) Start(ctx context.Context) error {
+	payload, err := tel.push(ctx, PayloadType.Init)
+	if err != nil {
+		tel.failedToStart = true
+		return fmt.Errorf("push: %w", err)
+	}
+	go func() {
+		t := time.NewTicker(tel.pushInterval)
+		for {
+			select {
+			case <-tel.shutdown:
+				return
+			case <-t.C:
+				payload, err = tel.push(ctx, PayloadType.Update)
+				if err != nil {
+					tel.logger.
+						WithField("action", "telemetry_push").
+						WithField("payload", fmt.Sprintf("%+v", payload)).
+						WithField("retry_at", time.Now().Add(tel.pushInterval).Format(time.RFC3339)).
+						Error(err.Error())
+					return
+				}
+				tel.logger.
+					WithField("action", "telemetry_push").
+					WithField("payload", fmt.Sprintf("%+v", payload)).
+					Info("telemetry update")
+			}
+		}
+	}()
+
+	tel.logger.
+		WithField("action", "telemetry_push").
+		WithField("payload", fmt.Sprintf("%+v", payload)).
+		Info("telemetry started")
+	return nil
+}
+
+// Stop shuts down the telemeter
+func (tel *Telemeter) Stop(ctx context.Context) error {
+	if tel.failedToStart {
+		return nil
+	}
+
+	waitForShutdown := make(chan struct{})
+	go func() {
+		tel.shutdown <- struct{}{}
+		waitForShutdown <- struct{}{}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("shutdown telemetry: %w", ctx.Err())
+	case <-waitForShutdown:
```

Comment:
if we wait just for `tel.shutdown <- struct{}{}` to be send i think we can put it in `default`instead of creating a channel and it will block until sent  or if we have a way to check it's status

Reply:
Is there a downside to the current approach?

Reply:
not really, i was just think it would be easier to read :)

Reply:
I am just thinking to be safer and not spawn a routine because it could make it side cornered to see issues, but the decision is yours ;)